lib/gc/Dialect/Linalgx/LinalgxOps.cpp

```diff
@@ -82,6 +82,385 @@ void SigmoidOp::getEffects(
+                         block.getArgument(0));
+  Value value2 =
+      helper.buildTypeFn(TypeFn::cast_signed, block.getArgument(2).getType(),
+                         block.getArgument(1));
+  Value value3 = helper.buildBinaryFn(BinaryFn::mul, value1, value2);
+  Value value4 =
+      helper.buildBinaryFn(BinaryFn::add, block.getArgument(2), value3);
+  yields.push_back(value4);
+  helper.yieldOutputs(yields);
+}
+
+ParseResult Mmt4DVnniOp::parse(OpAsmParser &parser, OperationState &result) {
+  return ::parseNamedStructuredOp(parser, result,
+                                  Mmt4DVnniOp::getNumRegionArgs(),
+                                  Mmt4DVnniOp::getRegionBuilder());
+}
+
+void Mmt4DVnniOp::print(OpAsmPrinter &p) {
+  ::printNamedStructuredOp(p, getOperation(), getInputs(), getOutputs());
+}
+
+LogicalResult Mmt4DVnniOp::fold(FoldAdaptor, SmallVectorImpl<OpFoldResult> &) {
+  return memref::foldMemRefCast(*this);
+}
+
+void Mmt4DVnniOp::getEffects(
+    SmallVectorImpl<SideEffects::EffectInstance<MemoryEffects::Effect>>
+        &effects) {
+  if (hasPureTensorSemantics())
+    return;
+  getGenericEffectsImpl(effects, getOperation()->getResults(), getDpsInputs(),
+                        getDpsInits());
+}
+
+LogicalResult Mmt4DVnniOp::verify() {
+  // A[M0, K0, M1, K]
+  // B[N0, K0, K1, N1, K2]
+  // C[M0, N0, M1, N1]
+  auto shapeA = ShapeAdaptor(getInputs()[0].getType());
+  auto shapeB = ShapeAdaptor(getInputs()[1].getType());
+  auto shapeC = ShapeAdaptor(getOutputs()[0].getType());
+  // check rank
+  auto hasRank = shapeA.hasRank() && shapeB.hasRank() && shapeC.hasRank();
+  if (!hasRank)
+    return emitOpError() << "input/output must have rank.";
+  auto checkRank = (shapeA.getRank() == 4) && (shapeB.getRank() == 5) &&
+                   (shapeC.getRank() == 4);
+  if (!checkRank)
+    return emitOpError() << "not supported input/output shape.";
+  // match M0, M1, N0, N1, K0, K dims
+  bool matchM0 = shapeA.getDimSize(0) == shapeC.getDimSize(0);
+  bool matchM1 = shapeA.getDimSize(2) == shapeC.getDimSize(2);
+  bool matchN0 = shapeB.getDimSize(0) == shapeC.getDimSize(1);
+  bool matchN1 = shapeB.getDimSize(3) == shapeC.getDimSize(3);
+  bool matchK0 = shapeA.getDimSize(1) == shapeB.getDimSize(1);
+  bool matchK =
+      shapeA.getDimSize(3) == (shapeB.getDimSize(2) * shapeB.getDimSize(4));
+  bool matchVnni = (shapeB.getDimSize(4) == 1) || (shapeB.getDimSize(4) == 2) ||
+                   (shapeB.getDimSize(4) == 4);
+  bool result = matchM0 && matchM1 && matchN0 && matchN1 && matchK0 && matchK &&
+                matchVnni;
+  if (!result)
+    return emitOpError() << "input/output dims packing not match.";
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// MultiBatchMatmulOp
+//===----------------------------------------------------------------------===//
+
+SmallVector<utils::IteratorType> MultiBatchMatmulOp::getIteratorTypesArray() {
```

Comment:
Shall we also add `verify` func to check the dims for this op?

Reply:
MultiBatchMatmulOp extends ContractionInterface, which use `mlir::linalg::detail::verifyStructuredOpInterface` and  `mlir::linalg::detail::verifyContractionInterface` to verify. I tested some mismatched dims, it check the dims as expected.

Reply:
I see, thanks.